Implemented change course role assignment.

waheedahmed opened the pull request:
[ECOM-6207](https://openedx.atlassian.net/browse/ECOM-6207)

**AC:**

1. Validate that all edX staff can change role assignments
2. Validate all future email notifications and ownership is transferred to the new user.

**Note:**

1. Doing this as an AJAX call is preferable, but be cognizant of how long it will take to implement.
2. ANY edX user can change ANY of the marketing, pc, and publisher role assignments.

**Implementation Details:**

1. Create group for all edX users
2. Check that user is a member of that group
3. DO NOT check for a particular permission for that user.

Prototype for changing permissions: http://s2q9k4.axshare.com/#g=1&p=ecom-6207&c=1

**Screenshot 1:**
<img width="325" alt="screen shot 2016-12-06 at 4 40 08 pm" src="https://cloud.githubusercontent.com/assets/2851134/20924338/c0d4e630-bbd2-11e6-8b67-c6a44680a5b1.png">

**Screenshot 2:**
<img width="325" alt="screen shot 2016-12-06 at 4 40 19 pm" src="https://cloud.githubusercontent.com/assets/2851134/20924339/c0d50200-bbd2-11e6-9ed6-31301fd43faa.png">

waheedahmed commented:
Base branch was merged into master, closed this PR https://github.com/edx/course-discovery/issues/455 and created this against master.

@bderusha Addressed your comments and added tests, please take another look.

@awais786 , @tasawernawaz , @ahsan-ul-haq Please review.

waheedahmed commented:
I'll update the translation later.

clintonb commented on course_discovery/apps/publisher/models.py:
```diff
@@ -388,3 +389,26 @@ def __str__(self):
+
+    class Meta:
+        unique_together = (
+            ('course', 'user', 'role'),
```
Should a user be able to have more than one role for a course run?

waheedahmed replied:
@clintonb No, one user can have only one role for a course.

clintonb replied:
In that case, role should be removed from your compound key.

waheedahmed replied:
You are right, thanks.

waheedahmed replied:
@clintonb We just discussed it with Deen and he said a user can have more than one role for a course, so I'm changing it back to ```('course', 'user', 'role')```.

clintonb commented on course_discovery/apps/publisher/tests/factories.py:
```diff
@@ -12,8 +12,8 @@
-    Course, CourseRun, OrganizationUserRole, Seat, State, UserAttributes
-)
+    Course, CourseRun, CourseUserRole, OrganizationUserRole, Seat, State,
+    UserAttributes)
```
Nit: add a newline

clintonb commented on course_discovery/apps/publisher/models.py:
```diff
@@ -369,6 +369,7 @@ class OrganizationUserRole(TimeStampedModel):
         (COORDINATOR, _('Partner Coordinator')),
         (REVIEWER, _('Reviewer')),
         (PUBLISHER, _('Publisher')),
+        (COURSE_TEAM, _('Course Team')),
```
Use `DjangoChoices`.  See `/course_discovery/apps/course_metadata/choices.py` and https://github.com/bigjason/django-choices.

clintonb commented on course_discovery/apps/publisher/views.py:
```diff
@@ -354,3 +382,27 @@ def post(self, request):
     serializer_class = UpdateCourseKeySerializer
+
+
+class CourseRoleAssignmentView(mixins.LoginRequiredMixin, View):
```
This looks an awful lot like an API endpoint. Why isn't it one?

clintonb commented on course_discovery/static/sass/publisher/publisher.scss:
```diff
@@ -637,3 +637,54 @@ select {
+
+        }
+    }
+}
```
Add a new line.

clintonb commented on course_discovery/templates/comments/comments_list.html:
```diff
@@ -1,7 +1,7 @@
 {% load i18n %}
```
Partials should be prefixed with an underscore (e.g. `_comments_list.html`).

bderusha replied:
@waheedahmed this still needs to be addressed

clintonb commented on course_discovery/templates/comments/add_auth_comments.html:
```diff
@@ -2,7 +2,7 @@
 {% load staticfiles %}
```
Partials should be prefixed with an underscore (e.g. `_add_auth_comments.html`).

bderusha replied:
@waheedahmed this still needs to be addressed

clintonb commented on course_discovery/templates/publisher/course_run_detail.html:
```diff
@@ -9,64 +9,64 @@
+        <nav class="administration-nav">
+            <div class="tab-container">
+                <button class="selected" data-tab="#tab-1">{% trans "All" %}</button>
+                <button data-tab="#tab-2">{% trans "STUDIO" %}</button>
```
Add translators comments explaining what these are. CAT, for example, is not the animal. It's an acronym for Course Administration Tool.

waheedahmed commented:
@clintonb Please review this PR.

clintonb commented on course_discovery/apps/publisher/permissions.py:
```diff
@@ -0,0 +1,25 @@
+        return check_role_assignment_permission(request.user)
+
+
+def check_role_assignment_permission(user):
```
This is not secure. This permission returns `True` for any user with any course or organization role. That would mean this permission *must always* be paired with another more restrictive class to be effective. Given that, this should either be a mixin or merged with the other permission class.

clintonb commented on course_discovery/static/js/publisher/views/course_detail.js:
```diff
@@ -0,0 +1,40 @@
+                console.log("Role assignment changed successfully.");
+            },
+            error: function (error) {
+                console.log("There was an error in changing role.");
```
Logging to the console does nothing for the user. Display a message to the user.

clintonb commented on course_discovery/static/js/publisher/views/course_detail.js:
```diff
@@ -0,0 +1,40 @@
+        var headers = {
+                "X-CSRFToken": Cookies.get("course_discovery_csrftoken")
+            },
+            apiEndpoint = $(this).data("apiEndpoint"),
```
`data('api-endpoint')`

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -1218,3 +1270,99 @@ def assert_email_sent(self, object_path, subject, expected_body):
+        user.groups.add(course_team_group)
+        assign_perm(Course.VIEW_PERMISSION, course_team_group, self.course)
+
+        self.client.logout()
```
This, and the lines below, should be refactored into an assertion method. This method can be used here and in the test above.

```python
def assert_user_cannot_change_course_role(self, user):
    """ Asserts the given user CANNOT change CourseUserRole objects. """
    self.client.logout()
    self.client.login(username=user.username, password=USER_PASSWORD)

    response = self.client.patch(
        self.change_assignment_url,
        data=json.dumps({'user': user.id}),
        content_type=JSON_CONTENT_TYPE
    )
    self.assertEqual(response.status_code, 403)

```

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -1218,3 +1270,99 @@ def assert_email_sent(self, object_path, subject, expected_body):
+
+        # Create a marketing reviewer group and assign two users.
+        reviewer_group = factories.GroupFactory(name=PublisherUserRole.MarketingReviewer)
+        self.user.groups.add(reviewer_group)
```
Name your users appropriately: `self.marketing_user`.

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -1218,3 +1270,99 @@ def assert_email_sent(self, object_path, subject, expected_body):
+        # Create a marketing reviewer group and assign two users.
+        reviewer_group = factories.GroupFactory(name=PublisherUserRole.MarketingReviewer)
+        self.user.groups.add(reviewer_group)
+        self.user2 = UserFactory()
```
`self.other_marketing_user`

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -1218,3 +1270,99 @@ def assert_email_sent(self, object_path, subject, expected_body):
+        assign_perm(Course.VIEW_PERMISSION, reviewer_group, self.course)
+
+        self.user3 = UserFactory()
+        # Create a coordinator group and assign it to user3.
```
Move this comment up, above the user creation.

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -1218,3 +1270,99 @@ def assert_email_sent(self, object_path, subject, expected_body):
+        self.user2.groups.add(reviewer_group)
+        assign_perm(Course.VIEW_PERMISSION, reviewer_group, self.course)
+
+        self.user3 = UserFactory()
```
`self.partner_coordinator`

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -1218,3 +1270,99 @@ def assert_email_sent(self, object_path, subject, expected_body):
+        organization = OrganizationFactory(name='testX')
+        self.course.organizations.add(organization)
+
+        self.user_default_role = factories.OrganizationUserRoleFactory(
```
This does not need to be an instance variable.

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -1218,3 +1270,99 @@ def assert_email_sent(self, object_path, subject, expected_body):
+        # verify that coordinator can't change reviewer role for this course.
+        self.assertEqual(response.status_code, 403)
+
+    def assert_role_assignment(self, response, user, expected_json):
```
Add a docstring explaining the purpose of this method.

clintonb replied:
Actually, this assertion method doesn't need to exist. It's only used in one location.

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -1218,3 +1270,99 @@ def assert_email_sent(self, object_path, subject, expected_body):
+
+
+class CourseRoleAssignmentViewTests(TestCase):
+    """ Tests for `CourseRoleAssignmentView` """
```
This docstring is not necessary.

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -1218,3 +1270,99 @@ def assert_email_sent(self, object_path, subject, expected_body):
+        self.client.login(username=user.username, password=USER_PASSWORD)
+        response = self.client.patch(
+            self.change_assignment_url,
+            data=json.dumps({'user': self.user2.id}),
```
`self.user2.id` can be any integer. You should not be referencing `self.user`.

waheedahmed commented:
@clintonb There are some requirement changes for this ticket, we added those on the ticket and also in the description of this PR. Kindly take another look, also addressed all your concerns.

waheedahmed commented:
@clintonb Please review this PR.

clintonb commented on course_discovery/apps/publisher/views.py:
```diff
@@ -83,10 +94,33 @@ class CourseRunDetailView(mixins.LoginRequiredMixin, mixins.ViewPermissionMixin,
-        context['comment_object'] = self.object.course
+
+        # initialize `CourseRunWrapper` on `CourseRun` object
+        course_run = CourseRunWrapper(context['object'])
```
You can clean this up by overriding `get_object()`. Also, consider setting [`context_object_name`](https://docs.djangoproject.com/en/1.10/ref/class-based-views/mixins-multiple-object/#django.views.generic.list.MultipleObjectMixin.context_object_name) to provide a more informative name.

waheedahmed replied:
Using ```get_object()``` here but setting ```context_object_name``` will be a big change, we will do that in cleanup.

clintonb commented on course_discovery/apps/publisher/views.py:
```diff
@@ -83,10 +94,33 @@ class CourseRunDetailView(mixins.LoginRequiredMixin, mixins.ViewPermissionMixin,
+
+        # Show role assignment widgets if user is an internal user.
+        if is_internal_user(self.request.user):
+            course_roles = course_run.course.course_user_roles.all().exclude(role=PublisherUserRole.CourseTeam)
```
I don't think you need `.all()` here.

clintonb commented on course_discovery/apps/publisher/views.py:
```diff
@@ -83,10 +94,33 @@ class CourseRunDetailView(mixins.LoginRequiredMixin, mixins.ViewPermissionMixin,
     model = CourseRun
     template_name = 'publisher/course_run_detail.html'
 
+    def get_role_widgets_data(self, course_roles):
```
Add a docstring explaining what's going on here.

waheedahmed replied:
Added.

clintonb commented on course_discovery/templates/publisher/course_run_detail.html:
```diff
@@ -9,64 +9,68 @@
+        <nav class="administration-nav">
+            <div class="tab-container">
+                <button class="selected" data-tab="#tab-1">{% trans "All" %}</button>
+                {% comment %}Translators: It's an edX tool to build courses{% endcomment %}
```
Studio is an edX tool for course creation.

clintonb commented on course_discovery/templates/publisher/course_run_detail.html:
```diff
@@ -9,64 +9,68 @@
+                <button class="selected" data-tab="#tab-1">{% trans "All" %}</button>
+                {% comment %}Translators: It's an edX tool to build courses{% endcomment %}
+                <button data-tab="#tab-2">{% trans "STUDIO" %}</button>
+                {% comment %}Translators: It's an acronym for Course Administration Tool{% endcomment %}
```
CAT is an acronym for Course Administration Tool.

clintonb commented on course_discovery/templates/publisher/course_run_detail.html:
```diff
@@ -9,64 +9,68 @@
+                {% comment %}Translators: It's an acronym for Course Administration Tool{% endcomment %}
+                <button data-tab="#tab-3">{% trans "CAT" %}</button>
+                {% comment %}Translators: It's an edX marketing site{% endcomment %}
+                <button data-tab="#tab-4">{% trans "DRUPAL" %}</button>
```
We should call this "Marketing Site".

waheedahmed replied:
This was provided in design prototype and on ticket.

clintonb commented on course_discovery/templates/publisher/course_run_detail.html:
```diff
@@ -9,64 +9,68 @@
+                {% comment %}Translators: It's an edX marketing site{% endcomment %}
+                <button data-tab="#tab-4">{% trans "DRUPAL" %}</button>
+                {% comment %}Translators: It's an Customer Relationship Management System{% endcomment %}
+                <button data-tab="#tab-5">{% trans "Salesforce" %}</button>
```
Salesforce is a brand name, and should not be translated.

clintonb commented on course_discovery/static/js/publisher/views/course_detail.js:
```diff
@@ -0,0 +1,36 @@
+
+$(document).ready(function() {
+    $(".btn-change-assignment").click(function(e){
+        var headers = {
```
Move `headers` outside this function. There is no need to read the cookie multiple times.

clintonb commented on course_discovery/static/js/publisher/views/course_detail.js:
```diff
@@ -0,0 +1,36 @@
+
+$(document).ready(function() {
+    $(".btn-change-assignment").click(function(e){
```
At some point we will write this down, but use single quotes throughout JavaScript files.

clintonb commented on course_discovery/templates/publisher/course_run_detail/_approval_widget.html:
```diff
@@ -0,0 +1,48 @@
+                <strong>{{ role_widget.heading }}</strong>
+            </span>
+            <div class="role-assignment-container">
+                <div id="user-role-container-{{ role_widget.user_course_role.role }}">
```
Use camelCase for ID attributes.

waheedahmed commented on course_discovery/apps/publisher/serializers.py:
```diff
@@ -37,3 +38,21 @@ def update(self, instance, validated_data):
         return instance
+
+
+class RoleAssignmentSerializer(serializers.ModelSerializer):
```
Created the requested ticket [ECOM-6574](https://openedx.atlassian.net/browse/ECOM-6574).

waheedahmed commented:
@clintonb Addressed all your comments, please take another look.

waheedahmed commented:
@clintonb Kindly take another look on this PR.

clintonb commented on course_discovery/apps/publisher/api/serializers.py:
```diff
@@ -0,0 +1,23 @@
+"""Publisher Api Serializers"""
```
Nit: API

clintonb commented on course_discovery/apps/publisher/api/tests/test_serializers.py:
```diff
@@ -0,0 +1,31 @@
+        }
+
+    def test_validation(self):
+        serializer = self.serializer_class(self.course_user_role)
```
`serializer = self.serializer_class(self.course_user_role, context={'context': self.request})`

clintonb commented on course_discovery/apps/publisher/api/tests/test_serializers.py:
```diff
@@ -0,0 +1,31 @@
+"""Tests Publisher Api Serializers."""
```
Nit: API

clintonb commented on course_discovery/apps/publisher/api/tests/test_views.py:
```diff
@@ -0,0 +1,95 @@
+            'user': self.internal_user.id,
+            'role': user_course_role.role
+        }
+        self.assertJSONEqual(str(response.content, encoding='utf8'), expected_json)
```
`self.assertDictEqual(response.data, expected)`

clintonb commented on course_discovery/apps/publisher/tests/test_serializers.py:
```diff
@@ -30,8 +30,8 @@ def test_validation(self):
         self.course_run.save()  # pylint: disable=no-member
         serializer = self.serializer_class(self.course_run)
         serializer.context['request'] = self.request
-        expected = serializer.validate(serializer.data)
```
This change is not necessary. The older version is clearer.

clintonb commented on course_discovery/apps/publisher/tests/test_views.py:
```diff
@@ -897,6 +903,53 @@ def test_get_course_return_none(self):
+        publisher_user = UserFactory()
+        internal_user_group = Group.objects.get(name=INTERNAL_USER_GROUP_NAME)
+
+        for user in [self.user, pc_user, marketing_user, publisher_user]:
```
`internal_user_group.user_set.add(*(self.user, pc_user, marketing_user, publisher_user))`
